controller/controlplane/controller.go

```diff
@@ -227,6 +235,55 @@ func (r *Reconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Resu
 		log.Debug(logger, "DataPlane not set, deployment for ControlPlane will remain dormant")
 	}
 
+	log.Trace(logger, "validating watch namespaces for the ControlPlane")
+	if err := validateWatchNamespaces(cp, r.WatchNamespaces); err != nil {
+		// TODO: Set status condition on the ControlPlane.
+		// https://github.com/Kong/kong-operator/issues/1975
+		log.Debug(logger, "watch namespaces validation failed", "error", err)
+		return ctrl.Result{}, err
+	}
+
+	log.Trace(logger, "validating WatchNamespaceGrants exist for the ControlPlane")
+	validatedWatchNamespaces, err := r.validateWatchNamespaceGrants(ctx, cp)
+	if err != nil { //nolint:gocritic
+		k8sutils.SetCondition(
+			k8sutils.NewConditionWithGeneration(
+				kcfgcontrolplane.ConditionTypeWatchNamespaceGrantValid,
+				metav1.ConditionFalse,
+				kcfgcontrolplane.ConditionReasonWatchNamespaceGrantInvalid,
+				fmt.Sprintf("WatchNamespaceGrant(s) are missing or invalid for the ControlPlane: %v", err),
+				cp.GetGeneration(),
+			),
+			cp,
+		)
+		// We do not return here as we want to proceed with reconciling the Deployment.
+		// This will prevent users using the ControlPlane with previous
+		// WatchNamespaces spec.
+		// We do not patch the status here either because that's done below.
+	} else if cp.Spec.WatchNamespaces != nil {
```

Comment:
The `if - else if - else` is a bit confusing. I think is may be clearer to restructure like this and add comments to talk about which situation is each branch processing
```go
if err != nil {
// Set conditions for the case validation of watched namespaces failed
} else {
// Set conditions for the case validation of watched namespaces passed
    if cp.Spec.WatchNamespaces != nil {
    // Set condition when CP has watched namespaces
    } else {
    // Set condition when CP does not have watchNamespaces, indicating that the CP watches all the namespaces
    }

}
```

Reply:
WDYT about fe3a7e3b86bedf70df316d1e899a896ca38b57f0 ?